sysboundary/ionosphere.h

```diff
@@ -148,20 +151,130 @@ namespace SBC {
+       * 
+       * / 1000 for kg m^-3 to g cm^-3
+       * * 100 for m to cm
+       * results in / 10 inside pow().
+       * 
+       * \param p particle species enum Particles::PROTON and ELECTRON
+       * \param E energy in keV
+       * \param h altitude index
+       */
+      inline Real fangNormalizedColumnMass(cint p, creal E, cuint h) {
+         Real y;
+         switch(p) {
+            case Particles::PROTON:
+               y = 7.5 / E * pow(atmosphere[h].density * atmosphere[h].scaleHeight / 1e-4 / 10, 0.9);
+               break;
+            case Particles::ELECTRON:
+               y = 2 / E * pow(atmosphere[h].density * atmosphere[h].scaleHeight / 6e-6 / 10, 0.7);
+               break;
+            default:
+               cerr << "Invalid precipitating species!" << endl;
+               abort();
+         }
+         return y;
+      }
+      
+      /*!< Energy dissipation f as a function of altitude and energy E, for
+       * - electrons Fang 2010 https://doi.org/10.1029/2010GL045406 equation (4)
+       * - protons Fang 2013 https://doi.org/10.1002/jgra.50484 equation (6)
+       * 
+       * Note that the i indices are taken -1 as we are 0-indexing in the arrays P_ij.
+       * 
+       * \param p particle species enum Particles::PROTON and ELECTRON
+       * \param E energy in keV
+       * \param h altitude index
+       */
+      inline Real fangEnergyDissipation(cint p, creal E, cint h) {
+         creal y = fangNormalizedColumnMass(p, E, h);
+         
+         cint offset = p*12; // 0 for proton, 12 for electron, see indexing in P_ij
+         return
+           fC(offset+0,E) * pow(y, fC(offset+1,E)) * exp(-fC(offset+ 2,E) * pow(y, fC(offset+ 3,E)))
+         + fC(offset+4,E) * pow(y, fC(offset+5,E)) * exp(-fC(offset+ 6,E) * pow(y, fC(offset+ 7,E)))
+         + fC(offset+8,E) * pow(y, fC(offset+9,E)) * exp(-fC(offset+10,E) * pow(y, fC(offset+11,E))); // this line returns zero for electrons, see P_ij
+      }
+      
+      Real totalIonizationFang(
+         cint n,
+         cint h
+      );
       
       // Hardcoded constants for calculating ion production table
       // TODO: Make these parameters?
       constexpr static int productionNumAccEnergies = 60;
       constexpr static int productionNumTemperatures = 60;
-      constexpr static int productionNumParticleEnergies = 100;
+// moved to be a parameter below       constexpr static int productionNumElectronEnergies = 100;
       constexpr static Real productionMinAccEnergy = 0.1; // keV
       constexpr static Real productionMaxAccEnergy = 100.; // keV
       constexpr static Real productionMinTemperature = 0.1; // keV
       constexpr static Real productionMaxTemperature = 100.; // keV
       constexpr static Real ion_electron_T_ratio = 4.; // TODO: Make this a parameter (and/or find value from kinetics)
-      // Ionoisation production table
+      
+      // These are parameters already
```

Comment:
In 5 years, somebody will read this comment and be very confused by it, and it doesn't help code understanding at all.. so I suggest to just drop it.

Reply:
Wilco.